ceno_zkvm/src/tables/ram.rs

```diff
@@ -22,9 +31,45 @@ impl DynVolatileRamTable for HeapTable {
     const V_LIMBS: usize = UINT_LIMBS;
     const ZERO_INIT: bool = true;
     const DESCENDING: bool = false;
+    const DYNAMIC_OFFSET: bool = true;
+
+    fn addr_expr<E: ExtensionField>(
+        cb: &mut CircuitBuilder<E>,
+        params: &ProgramParams,
+    ) -> Result<(Expression<E>, StructuralWitIn), CircuitBuilderError> {
+        let max_len = Self::max_len(params);
+        let addr = cb.create_structural_witin(
+            || "addr",
+            StructuralWitInType::EqualDistanceDynamicSequence {
+                max_len,
+                offset_instance_id: cb.query_heap_start_addr()?.0 as WitnessId,
+                length_instance_id: cb.query_heap_shard_len()?.0 as WitnessId,
+                multi_factor: WORD_SIZE,
+                descending: Self::DESCENDING,
+            },
+        );
+        Ok((addr.expr(), addr))
+    }
 
-    fn offset_addr(params: &ProgramParams) -> Addr {
-        params.platform.heap.start
+    fn max_len(params: &ProgramParams) -> usize {
+        let max_size = (params.platform.heap.end - params.platform.heap.start)
+            .div_ceil(WORD_SIZE as u32) as Addr;
+        1 << (u32::BITS - 1 - max_size.leading_zeros())
+    }
+
+    fn offset_addr(_params: &ProgramParams) -> Addr {
+        unimplemented!("heap offset is dynamic")
+    }
+
+    fn dynamic_offset_addr(params: &ProgramParams, pv: &PublicValues) -> Addr {
+        let heap_start = pv.heap_start_addr;
+        assert!(
+            heap_start >= params.platform.heap.start,
+            "heap_start {:x} < platform min heap start {:x}",
+            heap_start,
+            params.platform.heap.start
+        );
```

Comment:
we should also require that `heap_end <= params.platform.heap.end`.

Reply:
👍 done in https://github.com/scroll-tech/ceno/pull/1171/commits/00b1c712a59752f8eab1797206b76c5b21ccb0b8